corehq/apps/ota/management/commands/time_restore.py

```diff
@@ -0,0 +1,101 @@
+from django.core.management.base import BaseCommand
+from lxml import etree
+
+from corehq.apps.app_manager.dbaccessors import get_current_app_doc
+from corehq.apps.hqadmin.views.users import AdminRestoreView
+from corehq.apps.ota.views import get_restore_response
+from corehq.apps.users.models import CouchUser
+
+
+class Command(BaseCommand):
+    help = ("Runs a restore for the passed in user(s) and generates a csv with timing information. Example: "
+            "./manage.py time_restore test@ccqa.commcarehq.org,ethan@ccqa.commcarehq.org --app_id=123abc")
+
+    def add_arguments(self, parser):
+        parser.add_argument('domain')
+        parser.add_argument('usernames')
+        parser.add_argument('--app_id')
+
+    def handle(self, **options):
+        domain = options['domain']
+        usernames = options['usernames'].split(',')
+        app_id = options['app_id']
+        users = [CouchUser.get_by_username(username) for username in usernames]
+        for username in usernames:
+            if not CouchUser.get_by_username(username):
+                print("User '{}' not found".format(username))
+                return
+        if app_id:
+            try:
+                get_current_app_doc(domain, app_id)
+            except ResourceNotFound:
+                print("App '{}' not found".format(app_id))
+                return
+
+        headers, rows = _get_headers_and_rows(domain, users, app_id)
+        totals_row = _calculate_totals_row(headers, rows)
+
+        filename = "tmp.csv"
+        with open(filename, 'w') as f:
+            writer = csv.DictWriter(f, headers)
+            writer.writeheader()
+            writer.writerows(rows)
+            writer.writerow(totals_row)
+
+
+def _get_headers_and_rows(domain, users, app_id):
+    fixture_names = set()
+    rows = []
+    for user in users:
+        response, timing_context = get_restore_response(
+            domain, user, app_id=app_id,
+        )
+        timing_dict = timing_context.to_dict()
+        xml_payload = etree.fromstring(b''.join(response.streaming_content))
+        stats = AdminRestoreView.get_stats_from_xml(xml_payload)
+        row = {
+            'username': user.username,
+            'total_duration': timing_dict['duration'],
+            'num_cases': stats['num_cases'],
+            'num_locations': stats['num_locations'],
+            'num_reports': stats['num_reports'],
+            'num_ledger_entries': stats['num_ledger_entries'],
+        }
+        for provider in timing_dict['subs']:
+            row[provider['name']] = provider['duration']
+            if provider['name'] == 'FixtureElementProvider':
+                for fixture in provider['subs']:
+                    fixture_names.add(fixture['name'])
+                    row[fixture['name']] = fixture['duration']
+        rows.append(row)
+        print("Restore for '{}' took {} seconds to generate"
+              .format(user.username, timing_dict['duration']))
+
+    headers = [
```

Comment:
could be useful to add the time of the restore (as in date and time)

Reply:
:+1: